File: src/solar/solar-common/controller/BounceMissileController.ts
import Projectile from "@/Projectile";
import DamageType from "@/DamageType";
import WeaponType from "@/WeaponType";
import SelectUtil from "@/SelectUtil";
import UnitStateUtil from "@/UnitStateUtil";

/**
 * 弹跳弹道投射物(类似闪电链)
 */
export default class BounceMissileController implements Destroyable {
    unit: unit;
    range: number = 600
    /** 投射物 速度 */
    speed: number = 1000;
    loopCount: number = 5
    /** 已经击中过的单位 */
    hitUnitList: unit[] = []
    //伤害公式不为空的时候会自动造成公式算法的伤害
    damageStateFormula: StateFormula
    damageName?: string
    projectileModelPath: string
    /** 投射物特效模型大小缩放 */
    projectileModelScale: number = 1;
    /** 颜色rgb max=255 */
    colorR: number = 255
    colorG: number = 255
    colorB: number = 255
    /** 透明度 max=255 */
    alpha: number = 255
    damageTypeId: number = DamageType.T4_NORMAL;
    weaponTypeID: number = WeaponType.T0_WHOKNOWS;
    attackType: attacktype = ATTACK_TYPE_CHAOS
    /** 当更新时 */
    onUpdate: (controller: BounceMissileController) => void
    /** 当准备发射追踪投射物时 返回false取消发射 返回true继续*/
    onBeforeLaunchMissile: (controller: BounceMissileController, target: unit) => boolean
    /** 当发射追踪投射物时 可在这里修改具体的 */
    onLaunchMissile: (controller: BounceMissileController, projectile: Projectile) => void
    onHitTarget: (projectile: Projectile) => void
    //
    private _isDestroy: boolean = false;

    /** 任意附加数据 */
    [key: string]: any

    constructor(unit: unit, projectileModelPath: string = "Abilities\\Weapons\\SentinelMissile\\SentinelMissile.mdx", damageStateFormula?: StateFormula, onHitTarget?: (projectile: Projectile) => void) {
        this.unit = unit;
        this.projectileModelPath = projectileModelPath;
        this.damageStateFormula = damageStateFormula;
        this.onHitTarget = onHitTarget;
    }


    public launch(): Projectile {
        let x = GetUnitX(this.unit)
        let y = GetUnitY(this.unit)
        let lastHitUnit = null;
        if (this.hitUnitList.length > 0) {
            lastHitUnit = this.hitUnitList[this.hitUnitList.length - 1]
            x = GetUnitX(lastHitUnit);
            y = GetUnitY(lastHitUnit);
        }

        let enemyUnitList = SelectUtil.getEnemyUnitsInRangeOrderByDistance(this.unit, this.range, x, y);
        if (enemyUnitList == null || enemyUnitList.length == 0) {
            return
        }
        //
        if (enemyUnitList.length == 1 && enemyUnitList[0] == lastHitUnit) {
            //不能自己弹射自己 必须要有其他单位才行
            return;
        }

        //按距离 由近到远
        let targetUnit = null;
        for (let enemyUnit of enemyUnitList) {
            if (enemyUnit != lastHitUnit && !this.hitUnitList.includes(enemyUnit)) {
                //选择一个最近的新单位弹射
                targetUnit = enemyUnit;
                break;
            }
        }
        if (targetUnit == null) {
            //选择一个最近的已经弹射过的单位弹射
            for (let enemyUnit of enemyUnitList) {
                if (enemyUnit != lastHitUnit) {
                    //选择一个最近的新单位弹射
                    targetUnit = enemyUnit;
                    break;
                }
            }
        }
        if (targetUnit == null) {
            //正常不会走到这里 但是多判断一下总没有坏处 保守稳定大于0.1纳秒性能！
            return;
        }
        this.hitUnitList.push(targetUnit);
        //准备发射导弹
        if (this.onBeforeLaunchMissile) {
            let b = this.onBeforeLaunchMissile(this, targetUnit);
            //检查是否在监听里销毁了此类
            if (this._isDestroy) {
                return;
            }
            if (b == false) {
                return;
            }
        }
        //发射导弹
        let damage = 0
        if (this.damageStateFormula) {
            damage = UnitStateUtil.calculateStateFormula(this.damageStateFormula, this.unit, targetUnit)
        }
        let projectile: Projectile = new Projectile(this.unit, targetUnit, damage, this.projectileModelPath, this.onHitTarget);
        projectile.damageName = this.damageName
        projectile.speed = this.speed
        projectile.projectileModelScale = this.projectileModelScale
        projectile.colorR = this.colorR
        projectile.colorG = this.colorG
        projectile.colorB = this.colorB
        projectile.alpha = this.alpha
        projectile.damageTypeId = this.damageTypeId
        projectile.weaponTypeID = this.weaponTypeID
        projectile.attackType = this.attackType
        //手动设置发射点
        if (lastHitUnit != null) {
            projectile.x = GetUnitX(lastHitUnit)
            projectile.y = GetUnitY(lastHitUnit)
        }
        if (this.hitUnitList.length < this.loopCount) {
            //递归自己
            projectile.onHitTarget = p => {
                this.launch();
            }
        }
        //
        if (this.onLaunchMissile) {
            this.onLaunchMissile(this, projectile);
            //检查是否在监听里销毁了此类
            if (this._isDestroy) {
                return;
            }
        }
        projectile.start()
        return projectile
    }


    destroy() {
        this._isDestroy = true;
    }


}

File: scripts/solar/solar-common/controller/BounceMissileController.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ArrayIncludes = ____lualib.__TS__ArrayIncludes
local __TS__New = ____lualib.__TS__New
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["8"] = 1,["9"] = 1,["10"] = 2,["11"] = 2,["12"] = 3,["13"] = 3,["14"] = 4,["15"] = 4,["16"] = 5,["17"] = 5,["18"] = 10,["19"] = 10,["20"] = 10,["21"] = 46,["22"] = 46,["23"] = 46,["25"] = 12,["26"] = 14,["27"] = 15,["28"] = 17,["29"] = 23,["30"] = 25,["31"] = 26,["32"] = 27,["33"] = 29,["34"] = 30,["35"] = 31,["36"] = 32,["37"] = 41,["38"] = 47,["39"] = 48,["40"] = 49,["41"] = 50,["42"] = 46,["43"] = 54,["44"] = 55,["45"] = 56,["46"] = 57,["47"] = 58,["48"] = 59,["49"] = 60,["50"] = 61,["52"] = 64,["53"] = 65,["56"] = 69,["59"] = 75,["60"] = 76,["61"] = 77,["62"] = 79,["66"] = 83,["67"] = 85,["68"] = 86,["69"] = 88,["74"] = 93,["77"] = 97,["78"] = 97,["79"] = 99,["80"] = 100,["81"] = 102,["84"] = 105,["88"] = 110,["89"] = 111,["90"] = 112,["92"] = 114,["93"] = 114,["94"] = 114,["95"] = 114,["96"] = 114,["97"] = 114,["98"] = 114,["99"] = 114,["100"] = 115,["101"] = 116,["102"] = 117,["103"] = 118,["104"] = 119,["105"] = 120,["106"] = 121,["107"] = 122,["108"] = 123,["109"] = 124,["110"] = 126,["111"] = 127,["112"] = 128,["114"] = 130,["115"] = 132,["116"] = 133,["117"] = 132,["119"] = 137,["120"] = 138,["121"] = 140,["125"] = 144,["126"] = 145,["127"] = 54,["128"] = 149,["129"] = 150,["130"] = 149});
local ____exports = {}
local ____Projectile = require("solar.solar-common.tool.Projectile")
local Projectile = ____Projectile.default
local ____DamageType = require("solar.solar-common.constant.DamageType")
local DamageType = ____DamageType.default
local ____WeaponType = require("solar.solar-common.constant.WeaponType")
local WeaponType = ____WeaponType.default
local ____SelectUtil = require("solar.solar-common.util.unit.SelectUtil")
local SelectUtil = ____SelectUtil.default
local ____UnitStateUtil = require("solar.solar-common.util.unit.UnitStateUtil")
local UnitStateUtil = ____UnitStateUtil.default
____exports.default = __TS__Class()
local BounceMissileController = ____exports.default
BounceMissileController.name = "BounceMissileController"
function BounceMissileController.prototype.____constructor(self, unit, projectileModelPath, damageStateFormula, onHitTarget)
    if projectileModelPath == nil then
        projectileModelPath = "Abilities\\Weapons\\SentinelMissile\\SentinelMissile.mdx"
    end
    self.range = 600
    self.speed = 1000
    self.loopCount = 5
    self.hitUnitList = {}
    self.projectileModelScale = 1
    self.colorR = 255
    self.colorG = 255
    self.colorB = 255
    self.alpha = 255
    self.damageTypeId = DamageType.T4_NORMAL
    self.weaponTypeID = WeaponType.T0_WHOKNOWS
    self.attackType = ATTACK_TYPE_CHAOS
    self._isDestroy = false
    self.unit = unit
    self.projectileModelPath = projectileModelPath
    self.damageStateFormula = damageStateFormula
    self.onHitTarget = onHitTarget
end
function BounceMissileController.prototype.launch(self)
    local x = GetUnitX(self.unit)
    local y = GetUnitY(self.unit)
    local lastHitUnit = nil
    if #self.hitUnitList > 0 then
        lastHitUnit = self.hitUnitList[#self.hitUnitList]
        x = GetUnitX(lastHitUnit)
        y = GetUnitY(lastHitUnit)
    end
    local enemyUnitList = SelectUtil.getEnemyUnitsInRangeOrderByDistance(self.unit, self.range, x, y)
    if enemyUnitList == nil or #enemyUnitList == 0 then
        return
    end
    if #enemyUnitList == 1 and enemyUnitList[1] == lastHitUnit then
        return
    end
    local targetUnit = nil
    for ____, enemyUnit in ipairs(enemyUnitList) do
        if enemyUnit ~= lastHitUnit and not __TS__ArrayIncludes(self.hitUnitList, enemyUnit) then
            targetUnit = enemyUnit
            break
        end
    end
    if targetUnit == nil then
        for ____, enemyUnit in ipairs(enemyUnitList) do
            if enemyUnit ~= lastHitUnit then
                targetUnit = enemyUnit
                break
            end
        end
    end
    if targetUnit == nil then
        return
    end
    local ____self_hitUnitList_0 = self.hitUnitList
    ____self_hitUnitList_0[#____self_hitUnitList_0 + 1] = targetUnit
    if self.onBeforeLaunchMissile then
        local b = self:onBeforeLaunchMissile(self, targetUnit)
        if self._isDestroy then
            return
        end
        if b == false then
            return
        end
    end
    local damage = 0
    if self.damageStateFormula then
        damage = UnitStateUtil:calculateStateFormula(self.damageStateFormula, self.unit, targetUnit)
    end
    local projectile = __TS__New(
        Projectile,
        self.unit,
        targetUnit,
        damage,
        self.projectileModelPath,
        self.onHitTarget
    )
    projectile.damageName = self.damageName
    projectile.speed = self.speed
    projectile.projectileModelScale = self.projectileModelScale
    projectile.colorR = self.colorR
    projectile.colorG = self.colorG
    projectile.colorB = self.colorB
    projectile.alpha = self.alpha
    projectile.damageTypeId = self.damageTypeId
    projectile.weaponTypeID = self.weaponTypeID
    projectile.attackType = self.attackType
    if lastHitUnit ~= nil then
        projectile.x = GetUnitX(lastHitUnit)
        projectile.y = GetUnitY(lastHitUnit)
    end
    if #self.hitUnitList < self.loopCount then
        projectile.onHitTarget = function(____, p)
            self:launch()
        end
    end
    if self.onLaunchMissile then
        self:onLaunchMissile(self, projectile)
        if self._isDestroy then
            return
        end
    end
    projectile:start()
    return projectile
end
function BounceMissileController.prototype.destroy(self)
    self._isDestroy = true
end
return ____exports
